Add net_get_default_gateway() support for Apple platforms

past-due opened the pull request:
This is based on the Apple platform approach in https://github.com/miniupnp/libnatpmp/blob/master/getgateway.c (with IPv6-related changes - the IPv4 bits have been thoroughly tested to work).

Since that is BSD licensed, I've added it as separate file to maintain the separate licensing info. Would this be the intended way of including such code in this project?

paullouisageneau reviewed (commented):
Thank you for the PR, but I have an issue with the code license.

Even if the BSD 3-clause license is compatible with the MPL-2 license, it would require the distribution of any program compiled with the library to include the copyright notice. It would be cumbersome, and IMHO a short platform-specific function doesn't justify it.

past-due commented:
I've rewritten to use an adjusted approach, more in line with the behavior of the other platform `net_get_default_interface()` implementations, and included directly in net.c.

paullouisageneau commented on src/net.c:
```diff
@@ -205,6 +210,84 @@ int net_get_default_gateway(int family, addr_record_t *record) {
+		PLUM_LOG_WARN("sysctl[1] returned 0 buffer length");
+		return -1;
+	}
+	char *buf = (char*)malloc(buf_len);
```
You must check that the allocation did not fail. Also, there is no need to cast the return value.

paullouisageneau commented on src/net.c:
```diff
@@ -205,6 +210,84 @@ int net_get_default_gateway(int family, addr_record_t *record) {
+		}
+
+		if (family == AF_INET) {
+			if (((struct sockaddr_in *)sa_dst)->sin_addr.s_addr == 0) {
```
It would be better to compare the `INADDR_ANY` constant instead of the value 0.
```suggestion
			if (((struct sockaddr_in *)sa_dst)->sin_addr.s_addr == INADDR_ANY) {
```

paullouisageneau commented on src/net.c:
```diff
@@ -205,6 +210,84 @@ int net_get_default_gateway(int family, addr_record_t *record) {
+	int mib[NET_MIB_INTS] = {CTL_NET, PF_ROUTE, 0, family, NET_RT_FLAGS, RTF_GATEWAY};
+
+	size_t buf_len = 0;
+	if (sysctl(mib, NET_MIB_INTS, 0, &buf_len, 0, 0) != 0) {
```
```suggestion
	if (sysctl(mib, NET_MIB_INTS, NULL, &buf_len, NULL, 0) != 0) {
```

paullouisageneau commented on src/net.c:
```diff
@@ -205,6 +210,84 @@ int net_get_default_gateway(int family, addr_record_t *record) {
+		return -1;
+	}
+	char *buf = (char*)malloc(buf_len);
+	if (sysctl(mib, NET_MIB_INTS, buf, &buf_len, 0, 0) != 0) {
```
```suggestion
	if (sysctl(mib, NET_MIB_INTS, buf, &buf_len, NULL, 0) != 0) {
```

paullouisageneau reviewed (commented):
Thank you! I have a few comments, mostly nitpicks apart from the missing malloc check.

past-due commented:
Implemented your suggestions. Also added a check for link-local IPv6 gateway addresses, which helps filter some default macOS link-local results (that we do not want returned in any case).

paullouisageneau reviewed (approved):
It looks good to me, thank you for adding!
